stock_routing_operation/models/stock_move.py

```diff
@@ -0,0 +1,317 @@
+        return self + new_moves
+
+    def _apply_move_location_src_routing_operation(self):
+        """Apply routing operations
+
+        When a move has a routing operation configured on its location and the
+        destination of the move does not match the destination of the routing
+        operation, this method updates the move's destination and it's picking
+        type with the routing operation ones and creates a new chained move
+        after it.
+        """
+        for move in self:
+            if move.state not in (
+                "assigned",
+                "partially_available",
+            ) or move._bypass_routing_operation_application("src"):
+                continue
+
+            # Group move lines per source location, some may need an additional
+            # operations while others not. Store the number of products to
+            # take from each location, so we'll be able to split the move
+            # if needed.
+            # At this point, we should not have lines with different source
+            # locations, they have been split in
+            # _split_per_routing_operation(), so we can take the first one
+            source = move.move_line_ids[0].location_id
+            destination = move.move_line_ids[0].location_dest_id
+            # we have to add a move as destination
+            # we have to add move as origin
+            routing = source._find_picking_type_for_routing("src")
+            if not routing:
+                continue
+
+            if self.env["stock.location"].search(
+                [
+                    ("id", "=", routing.default_location_dest_id.id),
+                    ("id", "parent_of", move.location_dest_id.id),
+                ]
+            ):
+                # we don't need to do anything because we already go through
+                # the expected destination
+                continue
+
+            # the current move becomes the routing move, and we'll add a new
+            # move after this one to pick the goods where the routing moved
+            # them, we have to unreserve and assign at the end to have the move
+            # lines go to the correct destination
+            move._do_unreserve()
+            move.package_level_id.unlink()
+            dest = routing.default_location_dest_id
+            current_picking_type = move.picking_id.picking_type_id
+            move.write({"location_dest_id": dest.id, "picking_type_id": routing.id})
+            move._insert_routing_moves(
+                current_picking_type, move.location_id, destination
+            )
+
+            picking = move.picking_id
+            move._assign_picking()
+            if not picking.move_lines:
+                # When the picking type changes, it will create a new picking
+                # for the move. If the previous picking has no other move,
+                # we have to drop it.
+                picking.unlink()
```

Comment:
Hi @guewen 
I know everything in the module may change drastically, but this problematic will probably continues (if this is done at reservation time at least), so I share my thoughts already.
I think deleting the picking is important, because as you say, it may be empty, but doing it here is an issue.

Indeed, the `_action_assign `of the `stock.move` is normally called by `action_assign` on `stock.picking`
And, in case we call `action_assign` on the pickings, the risk is that the picking itself is deleted during the call.
Meaning that any action on the picking done after the call of `moves._action_assign()` may fail.

I am not sure it happens with Odoo natives modules only, but I guess it is a big risk to conflict with multiple OCA modules or custom modules I guess.
Any module overriding `action_assign` of `stock.picking` and doing stuff after the super has big chances to fail because the record won't exist anymore.

Well, Unfortunately, I don't really have a good solution about it... I guess we could delete pickings afterward with a cron...(or a job but we don't want to depend on queue job for this module of course...)

What do you think?
Another minor issue is that if a user click on check availability on the button present in the picking form, it may be deleted. It may be a bit disturbing.

Reply:
> Well, Unfortunately, I don't really have a good solution about it... I guess we could delete pickings afterward with a cron...(or a job but we don't want to depend on queue job for this module of course...)

And it wouldn't solve anything, because if a module *does something* with the picking, the *does something* part is probably lost, or doing unwanted behavior.
@jgrandguillaume thoughts?

Reply:
> And maybe it wouldn't solve everything, because if a module does something with the picking, the does something part is probably lost, or doing unwanted behavior.

I agree, it would not solve everything. Something may apply for instance to the new created picking (the replacement(s)).
At least it would not fail.

Another issue with the change of picking is that some relevant information present in the original picking, may be relevant (or not) for the newly created picking, but the `_assign_picking` won't take care of it. Because it just put some basic info, but it is not a "copy" of the initial picking.
Like a data manually put on the initial picking, that would still be relevant for the new picking replacing it. (I don't specially have an exemple, could be a simple note relative to the expedition...)
In the same way, depending on the information/case, we would want or not to manage it...
Well, I guess this may be a specific part to manage if necessary that do not need to be attended in the module.

Reply:
we can cancel the picking, a bit better than deleting

Reply:
Yes, I think it is a good solution. 

I would do this part in a separated method, in which we cancel the picking.
like `picking.post_process...()`
So anyone could override it easily to do another processing for this kind of picking.  Like if I want to delete it because I know it is safe depending on the other module installed on my instance, I still can do it. (or just delay the deletion).

Ok for me to merge then!

Reply:
Not ready for merge though :D

Reply:
What is missing? (I mean beside this post_process matter)

Reply:
As I said earlier, I have a lot of things to change in this module, notably the datamodel will have dedicated models. 
The change I'm working on currently is a change of the rules:

* if the move's dest location is a child of the routing's dest location: it's more precise so change only the picking type
* if the move's dest location is a parent of the routing's dest location: change the dest location to the routing's dest location and change the picking type
* if the move's dest location is outside of the routing's dest location: add a routing operation before

Which means that even if your destinations were the right ones, the move will be classified in the expected picking type.

Reply:
> Which means that even if your destinations were the right ones, the move will be classified in the expected picking type.

That's nice, that is something I wanted to change. (change at least picking type, even if location dest is the same for instance)